crates/compiler/codegen/src/optimizations.rs

```diff
@@ -0,0 +1,472 @@
+//! # CASM Optimizations
+//!
+//! This module contains optimization passes for CASM instructions.
+
+use std::collections::{HashMap, HashSet};
+
+use crate::builder::SymbolicInstruction;
+use crate::{CasmBuilder, CasmInstruction, CodegenResult, Opcode};
+
+/// Trait for CASM optimization passes
+pub trait CasmOptimization {
+    /// Name of the optimization pass
+    fn name(&self) -> &'static str;
+
+    /// Run the optimization pass on a function's instructions
+    fn optimize(&self, builder: &mut CasmBuilder) -> CodegenResult<()>;
+}
+
+/// Collection of optimization passes
+pub struct OptimizationPipeline {
+    passes: Vec<Box<dyn CasmOptimization>>,
+}
+
+impl OptimizationPipeline {
+    /// Create a new optimization pipeline
+    pub fn new() -> Self {
+        Self { passes: Vec::new() }
+    }
+
+    /// Add an optimization pass to the pipeline
+    pub fn add_pass(&mut self, pass: Box<dyn CasmOptimization>) {
+        self.passes.push(pass);
+    }
+
+    /// Run all optimization passes in sequence
+    pub fn run(&self, builder: &mut CasmBuilder) -> CodegenResult<()> {
+        for pass in &self.passes {
+            pass.optimize(builder)?;
+        }
+        Ok(())
+    }
+}
+
+impl Default for OptimizationPipeline {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Neutral operations pass
+///
+/// This pass eliminates neutral operations (add 0, sub 0, mul 1, div 1)
```

Comment:
these operations should very likely be eliminated at the MIR level, as they are not related to CASM at all and are just identity operations - so it would be way easier to detect them in MIR.